image_view/src/image_saver.cpp

```diff
@@ -0,0 +1,284 @@
+  rclcpp::Service<std_srvs::srv::Trigger>::SharedPtr start_srv_;
+  rclcpp::Service<std_srvs::srv::Trigger>::SharedPtr end_srv_;
+
+  bool saveImage(const sensor_msgs::msg::Image::ConstSharedPtr & image_msg, std::string & filename)
+  {
+    cv::Mat image;
+    try {
+      image = cv_bridge::toCvShare(image_msg, encoding)->image;
+    } catch(cv_bridge::Exception) {
+      RCLCPP_ERROR(
+        this->get_logger(), "Unable to convert %s image to %s",
+        image_msg->encoding.c_str(), encoding.c_str());
+      return false;
+    }
+
+    if (!image.empty()) {
+      try {
+        filename = (g_format).str();
+      } catch (...) {
+        g_format.clear();
+      }
+
+      try {
+        filename = (g_format % count_).str();
+      } catch (...) {
+        g_format.clear();
+      }
+
+      try { 
+        filename = (g_format % count_ % "jpg").str();
+      } catch (...) {
+        g_format.clear();
+      }
+
+      if (save_all_image || save_image_service ) {
+        cv::imwrite(filename, image);
+        RCLCPP_INFO(this->get_logger(), "Saved image %s", filename.c_str());
+
+        save_image_service = false;
+      } else {
+        return false;
+      }
+    } else {
+      RCLCPP_WARN(this->get_logger(), "Couldn't save image, no data!");
+      return false;
+    }
+
+    return true;
+  }
+
+public:
+  ImageSaverNode(const rclcpp::NodeOptions & options)
+    : rclcpp::Node("image_saver_node", options),
+      is_first_image_(true), has_camera_info_(false), count_(0)
+  {
+    auto topic = rclcpp::expand_topic_or_service_name("image", this->get_name(), this->get_namespace());
+
+    // Useful when CameraInfo is being published
+    cam_sub_ = image_transport::create_camera_subscription(
+      this, topic, std::bind(
+        &ImageSaverNode::callbackWithCameraInfo, this, std::placeholders::_1, std::placeholders::_2),
+      "raw");
+
+    // Useful when CameraInfo is not being published
+    image_sub_ = image_transport::create_subscription(
+      this, topic, std::bind(
+        &ImageSaverNode::callbackWithoutCameraInfo, this, std::placeholders::_1),
+      "raw");
+
+    std::string format_string;
+    format_string = this->declare_parameter("filename_format", std::string("left%04i.%s"));
+    encoding = this->declare_parameter("encoding", std::string("bgr8"));
+    save_all_image = this->declare_parameter("save_all_image", true);
+    request_start_end = this->declare_parameter("request_start_end", false);
+    g_format.parse(format_string);
+
+    if (request_start_end && !save_all_image) {
+      RCLCPP_WARN(
+        this->get_logger(), "'request_start_end' is true, so overwriting 'save_all_image' as true");
+    }
+
+    save_srv_ = this->create_service<std_srvs::srv::Empty>(
+      "save", std::bind(&ImageSaverNode::service, this, _1, _2, _3));
+
+    // FIXME(unkown): This does not make services appear
+    // if (request_start_end) {
+    start_srv_ = this->create_service<std_srvs::srv::Trigger>(
```

Comment:
Does this work while processing an image? I’d think with a single threaded executor this wouldn’t get called

Reply:
*While* an image is being processed, this will not be handled but it will go into the callback queue which will get handled on the next `spin()`. The queue is asynchronous but the processing of callbacks is sequential unless you use a MultiThreadedExecutor (in which case you have to worry more about synchronization). Even so, it wouldn't take effect until the next image is received and that image callback handled.